napari/util/keybindings.py

```diff
@@ -275,98 +284,135 @@ class KeybindingDescriptor:
+        super().__init__()
+        self.keymap = {}
 
-    def __getitem__(self, key):
-        try:
-            val = super().__getitem__(key)
-        except KeyError:
-            val = self._parent()[key]
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
 
-        if val is None:
-            raise KeyError(key)
+        if 'class_keymap' not in cls.__dict__:
+            # if in __dict__, was defined in class and not inherited
+            cls.class_keymap = {}
 
-        return val
+    bind_key = KeybindingDescriptor(bind_key)
 
-    def __contains__(self, key):
-        try:
-            self[key]
-        except KeyError:
-            return False
-        else:
-            return True
 
-    def __delitem__(self, key):
-        self[key] = None
+class KeymapHandler:
+    """Mix-in to add key handling functionality.
 
-    def copy(self):
-        return InheritedKeymap(self._parent, self.data.copy())
+    Attributes
+    ----------
+    keymap_providers : list of KeymapProvider
+        Classes that provide the keymaps for this class to handle.
+    """
 
-    def pop(self, key, default=UNDEFINED):
-        try:
-            default = self[key]
-        except KeyError:
-            if default is UNDEFINED:
-                raise
-        else:
-            del self[key]
+    def __init__(self):
+        super().__init__()
+        self._key_release_generators = {}
+        self.keymap_providers = []
 
-        return default
+    @property
+    def keymap_chain(self):
+        """collections.ChainMap: Chain of keymaps from keymap providers."""
+        maps = []
 
+        for parent in self.keymap_providers:
+            maps.append(parent.keymap)
+            maps.append(parent.class_keymap)
 
-class KeymapMixin:
-    """Mix-in to add keymap functionality. Must still define ``class_keymap``
-    in every subclass.
-    """
+        return ChainMap(*maps)
 
-    def __init__(self):
-        self.keymap = {}
+    @property
+    def active_keymap(self):
+        """dict: Active keymap, created by resolving the keymap chain."""
+        keymaps = self.keymap_chain.maps
 
-    def __init_subclass__(cls, **kwargs):
-        super().__init_subclass__(**kwargs)
+        for i, keymap in enumerate(keymaps):
+            if Ellipsis in keymap:  # catch-all
+                break
+        keymaps = keymaps[: i + 1]  # trim maps after a catch-all
 
-        if 'class_keymap' not in cls.__dict__:
-            # if in __dict__, was defined in class and not inherited
-            cls.class_keymap = {}
+        active_keymap = {}
 
-    @property
-    def keymap(self):
-        """InheritedKeymap : Keymap used for shortcuts.
-        Inherits from ``class_keymap``.
+        for keymap in reversed(keymaps):
+            for key, func in keymap.items():
+                active_keymap[key] = func
+                if func is Ellipsis:  # blocker
+                    del active_keymap[key]
```

Comment:
how about `active_keymap = dict(filter(lambda kv: kv[1] is not Ellipsis, keymaps.items()))`?

Reply:
Eh. I tend to prefer comprehensions to filters unless they're part of a pipe, but I'm ok with that, with one modification:

```python
dict(filter(lambda k, func: func is not Ellipsis, keymaps.items()))
```

But the three closing parentheses make this harder to read than my dict comprehension, I think... But again, I don't mind it. =)

Reply:
I wish I could do this but `dict.items` returns a list of tuples. And yeah you're right, it is harder to read, so I'll do the comprehension.

I also just realized instead of the reducing expression, we could use `dict(ChainMap(*keymaps))`, is this any more or less readable? I'm on the fence.

Reply:
> but dict.items returns a list of tuples

Oh right. 🤦‍♂ Well then I think the comprehension is nicer.

> is this any more or less readable?

Heh, nice. Despite the fun I had coming up with that reduction, I think using the chainmap is probably more readable, at least once you've encountered chainmaps a few times, but I'll let @sofroniewn comment.

Reply:
I'm not super familiar with chainmap, so defer to you both, but it's fine with me

Reply:
@sofroniewn but do you find the `reduce` call understandable or arcane? If you know that `keymaps` is a list of dictionaries?